pkg/elf/elf_test.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !privileged_tests
+
+package elf
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	. "gopkg.in/check.v1"
+)
+
+// Hook up gocheck into the "go test" runner.
+type ELFTestSuite struct{}
+
+var (
+	_ = Suite(&ELFTestSuite{})
+
+	elfObjCopy = "elf-demo-copy.o"
+)
+
+func Test(t *testing.T) {
+	TestingT(t)
+}
+
+func hash(path string) ([]byte, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	hash := sha256.New()
+	if _, err := io.Copy(hash, f); err != nil {
+		return nil, err
+	}
+
+	return hash.Sum(nil), nil
+}
+
+var errDifferentFiles = fmt.Errorf("File hashes differ")
+
+func compareFiles(path1, path2 string) error {
+	h1, err := hash(path1)
+	if err != nil {
+		return err
+	}
+	h2, err := hash(path2)
+	if err != nil {
+		return err
+	}
+
+	if len(h1) != len(h2) {
+		return fmt.Errorf("File hashes differ in length")
+	}
+
+	for i := range h1 {
+		if h1[i] != h2[i] {
+			return errDifferentFiles
+		}
+	}
+
+	return nil
+}
+
+func getDemoPath() (string, error) {
+	wd, err := os.Getwd()
+	if err != nil {
+		return "", fmt.Errorf("Failed to get working directory: %s", err)
+	}
+
+	return filepath.Join(wd, "..", "..", "test", "bpf", "elf-demo.o"), nil
```

Comment:
can you guarantee it will always be this relative path? Isn't there a way that we could pass the elf-demo.o location via a global variable or something like it? :thinking:

Reply:
Yes, I can guarantee it - the unit tests will fail otherwise ;)

In all seriousness though, I found a nice little Makefile/ldflags hack for this, I'll roll in an extra patch for this.